src/front/glsl/parser/declarations.rs

```diff
@@ -199,8 +199,31 @@ impl<'source> ParsingContext<'source> {
             // TODO: Should we try to make constants here?
             // This is mostly a hack because we don't yet support adding
             // bodies to entry points for variable initialization
-            let maybe_constant =
-                init.and_then(|(root, meta)| parser.solve_constant(ctx.ctx, root, meta).ok());
+
+            // Initializers for external declarations must be constants,
+            // so if `solve_constant` failed, that's an error the user should see.
+            // But if it's an initializer for a local variable, then failing to
+            // produce a constant just means we emit the expression as code;
+            // that is not an error.
+            let maybe_constant = if let Some((root, meta)) = init {
+                if ctx.qualifiers.is_empty() {
+                    None
+                } else {
+                    let &(ref qualifier, _meta) = &ctx.qualifiers[0];
+                    match *qualifier {
+                        TypeQualifier::StorageQualifier(StorageQualifier::Const) => {
+                            match parser.solve_constant(ctx.ctx, root, meta) {
+                                Ok(res) => Some(res),
+                                Err(err) if ctx.external => return Err(err),
+                                _ => None,
+                            }
+                        }
+                        _ => None,
+                    }
```

Comment:
This won't work because the constant qualifier isn't required to be the first try something like:
```rust
let is_const = ctx.qualifiers.iter().find(|(qualifier,_)| match qualifier {
  TypeQualifier::StorageQualifier(StorageQualifier::Const) => true,
  _ => false
}).is_some();
```